### PR TITLE
Create Translator Release Process Details

### PR DESCRIPTION
**Pull request by tursynay**
@NCATSTranslator/tact-voting-members, these are Translator Release Process Details that outline the step by step release order.
Please review and comment by the end of November 14. We will merge then.

### CODE REVIEW

**Comment by saramsey** [comment]
L46 addresses hotfix deployments to PROD during the coding window, but doesn't specifically address hotfix deployments to TEST during the coding window. What are the requirements to be able to deploy a hotfix to TEST during the coding window? Are they different than the requirements to be able to deploy a hotfix to PROD? I could see an argument for being more permissive for a hotfix to be deployed only to TEST during the coding window (i.e., to fix a bug that would otherwise make it difficult for SMEs to evaluate the new release candidate for Translator).

**Comment by sierra-moxon** [comment]
@saramsey - can I clarify your comment in my head? :) I think you might be asking two questions, but I am not sure:  

1) is the process for push to PROD during the "code window" the same as the process for push to TEST?   This is the phase in our cycle where TEST is a mirror of PROD.

My vote here would be yes.  We can use the same templated changelog as a request for pushes to TEST as we do for the same issue being pushed to PROD.  

2) is the process for pushes to TEST during the "freeze/slush and plan window" the same as during the "code window"?   This is the phase in our cycle when TEST becomes the site where we manually test the new release. 

My vote here would be no.  Once we get to the "slush and plan window", there will be lots of small bug fixes going into TEST from all the teams as testers manually find issues with the new release (to fix a bug that would otherwise make it difficult for SMEs to evaluate the new release candidate for Translator).  

We have not identified a process for deciding which things identified in the "slush and plan window" (via manual testing) to fix before release.  I would vote to keep the process very light in this phase; our collective "best judgment" as component developers could be the baseline.

**Comment by saramsey** [comment]
Thank you @sierra-moxon. That's very helpful.

I do have a follow-up question. I'm a bit confused by the statement that we would want TEST and PROD to be mirrors of one another during the coding window. Suppose that during the coding window, we (Team Expander Agent) propose to deploy a hot-fix for ARAX to address a high-severity bug that for whatever reason previously wasn't caught earlier in the Translator release process. I assume we would solicit and obtain approval from TACT and ITRB to deploy the patch to ARAX in TEST. Then we or others would _verify that the patch is working in TEST,_ before there would be any discussion of deploying it to PROD. Right? So there would be a period of time (perhaps short) in which TEST and PROD would not be mirrors; TEST would have the patch and PROD would not. Perhaps it's pedantic, but during that window, the two deployment environments would not be mirrors, by my way of thinking. (perhaps I misunderstand).

As for having a lightweight process for deploying fixes during the slush/plan window, that's fine with me. I presume deployment requests during that window would be limited to bug fixes, just as they are in the coding window. 

In summary, to see if I understand it right-- Translator is moving to a model where for new features and "performance enhancements", we only do deployments four times a year, during official releases (right?).  Outside of those four deployments at specific quarterly intervals, it's only bugfixes, and only with approvel -- lightweight approval if it is a SME/TAQA affecting issue during the slush window, and more heavyweight approval during the coding window. 

Do I have that right?

**Comment by sierra-moxon** [comment]
@saramsey - I think this is right - let me echo it back to you with slightly different terms to confirm :) : 

_I'm a bit confused by the statement that we would want TEST and PROD to be mirrors of one another during the coding window._

I should have been more specific: TEST and PROD should be mirrors of one another in the code window except when we do a hotfix.  Then TEST will be the place where we verify that the patch is working so that it can be deployed to PROD.  I imagine this turn-around cycle to be short and once the hotfix is released, TEST and PROD again are mirrors of each other.

_Translator is moving to a model where for new features and performance enhancements, we only do deployments four times a year, during official releases (right?)_

This is our current proposal, right - four planned releases each year where new features, fixes, and planned enhancements are deployed.  But, I want to reiterate that the timing of these releases can and should iterate -- we need to get through one cycle to see if it is too long or too short or needs to be completely refactored based on feedback from the teams.  

 _Outside of those four deployments at specific quarterly intervals, it's only hotfixes, and only with approval -- lightweight approval if it is a SME/TAQA affecting issue during the slush window, and more heavyweight approval during the coding window._

yep!  this is what I think our proposal is as well.  :)

**Comment by edeutsch** [comment]
I'm a bit confused how this is different from #4

File: Translator Release Process Details [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+Code Window
+Freeze and Prep Window
+Retrospective
+
+Code window (2 months): 
+GOAL: To establish a predictable schedule for deployment requests, coordinate deployments between interdependent components, and help improve our various teams’ ability to plan ahead.
+Definition: Do good work and keep in touch. 
+
+Freeze-and-prep window (1 month):
+GOAL: To come together as a unified consortia to coordinate agendas between our working groups and committees, be transparent about when work is coming for different teams, give testers the ability to see what they should be testing with priority for a release, and try to reduce the “pre-relay scramble.”
```
Comment (nlharris) [inline Translator Release Process Details]:
```suggestion
GOAL: To come together as a unified consortium to coordinate agendas between our working groups and committees, be transparent about when work is coming for different teams, give testers the ability to see what they should be testing with priority for a release, and try to reduce the “pre-relay scramble.”  
```

File: Translator Release Process Details [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+The SMuRF coordinator should be able to tell from the milestones identified for the release, what the major areas of work should be tested for the release, what components or MVPs or displays need to be tested/confirmed in TEST.
+Note: when more automated tests are in place, we can re-evaluate the kinds of manual testing required.
+SMuRF coordinator might want to keep a simple list of manual tests that are helpful to confirm every release. 
+TAQA will meet each week to help triage resulting issues back to teams, request feedback from the group about what constitutes a bug that must be fixed before release, and also keep track of new issues/features for prioritization in the next release.
+TACT decides on a case-by-case basis if production deployment is go/no-go for new component(s)
+TACT votes via procedures outlined in the current Consortium Guidelines - see Change Management section here
```
Comment (nlharris) [inline Translator Release Process Details]: "here" is not hyperlinked -- either add the link or remove the word "here"

File: Translator Release Process Details [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+Subsequent Release Planning
+Team/Working Group/Committee/External Group bring proposal(s) for new component(s) or extensive updates to existing component(s) to TACT for review
+Teams to include the following details:
```
Comment (nlharris) [inline Translator Release Process Details]:
```suggestion
The following details should be included in these proposals:
```

File: Translator Release Process Details [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+Team/Working Group/Committee/External Group bring proposal(s) for new component(s) or extensive updates to existing component(s) to TACT for review
+Teams to include the following details: 
+Theme (required) 
+e.g. what working group, committee, architecture component this task belongs to.
+Milestone: a short, executive summary of the proposed work. (required)
+e.g. - Fix Biolink Documentation: Deliver a refactored repository following LinkML best practices with documentation that is searchable and navigable as it previously was.
+e.g. - Host a series of discussions around modeling “variants” for use in Translator: At the end of the cycle, propose options to the Data Modeling team for confirmation. 
+e.g. - Revisit SRI confidence score [g() function]: deploy a revised scoring algorithm that takes SRI confidence score into account.
+Specific steps to accomplish the milestone (or a link to a GitHub project tracker) (required)
+Timeline for completing the proposed work for completing the proposed work
```
Comment (nlharris) [inline Translator Release Process Details]:
```suggestion
Timeline for completing the proposed work
```

File: Translator Release Process Details [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+Responsible teams (or individuals) (required)
+Status (optional)
+Any additional notes that help clarify the tasks represented by the deliverable (optional)
+e.g. - will this involved release coordination with any other teams?
```
Comment (nlharris) [inline Translator Release Process Details]:
```suggestion
e.g. - will this involve release coordination with any other teams? 
```

File: Translator Release Process Details [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+Review process, revise if necessary, and repeat as part of next release
+GOAL: formalize the need to not institutionalize processes that make production of Translator components unnecessarily difficult, slow, confusing, or monotonous.
+Retrospective participants should feel comfortable expressing their successes and challenges in the context of the release, bringing with them specific examples (when possible) of things that were good or bad, preferably with suggestions on how to make the process better.  
+Hotfixes
+GOAL: Allow changes to go into PROD during the coding window when they address critical errors in the deployed feature.
+Definition: a hotfix is a code change to address a bug that is deployed between official release cycles. Hotfixes do not address non-critical issues, new features, or incremental improvements and optimizations.
+Policy: Teams can propose hotfixes to be reviewed and approved by TACT
+TACT members making hotfix decisions, will greatly benefit from hotfix descriptions following a basic template:
+Sample Changelog for Code Window deployments to PROD
```
Comment (nlharris) [inline Translator Release Process Details]: I guess this and the next two items are awaiting content?

File: Translator Release Process Details [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+sample changelog for PROD deployment request - ARAX
+
+Translator Release Schedule Timeline
+Translator Release Process Guidelines
+
+
+Nomenclature:
+Code Window - develop the priorities set for the window
```
Comment (nlharris) [inline Translator Release Process Details]: I don't understand this definition of "Code Window"

File: Translator Release Process Details [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+Policy: Teams can propose hotfixes to be reviewed and approved by TACT
+TACT members making hotfix decisions, will greatly benefit from hotfix descriptions following a basic template:
+Sample Changelog for Code Window deployments to PROD
+sample changelog for PROD deployment request - ARAX
+
+Translator Release Schedule Timeline
+Translator Release Process Guidelines
+
+
+Nomenclature:
```
Comment (nlharris) [inline Translator Release Process Details]: The heading "Nomenclature" makes me think that what follows will be a glossary, but instead it seems like some residual internal notes

**Review by nlharris — COMMENTED** [review]
I'm pretty sure the formatting here will need some work (e.g., there's no separation between sections; things that are presumably bullet points or at least separate paragraphs are all run together; etc.) but it's hard to suggest specific improvements until I can see how this document looks once it's merged.

**Comment by karafecho** [comment]
**Note to all:** 

If approved, the Translator Release **Guidelines** doc will be archived with the other Translator Guidelines as a G-doc. Here is a [link](https://docs.google.com/document/d/1h4_UKf4gwnOMFyAYPM7WajZ43NHgHZX93eHHK3pzMtU/edit?usp=sharing) to the current draft of the guidelines in human-readable non-md format.

The Translator Release **Process Details** doc contains detailed steps regarding the release process that are not appropriate for the high-level guidelines doc, but rather are intended to inform Translator team members on how things are expected to work in practice. Sort of like an SOP.

Apologies for the confusion!

**Comment by tursynay** [comment]
During the winter 2024 release schedule retrospectives relay session, this process will be updated and finalized. Will merge this PR then.
